### PR TITLE
Add recipe for knowledge distillation

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
# Description

This PR creates a recipe for knowledge distillation.

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests
Documentation

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: getting_started/Knowledge_Distillation.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+--dataset-path HuggingFaceH4/ultrachat_200k --data-split train_sft \
+--data-columns messages \
+--max-prefill-length 64 --max-target-length 2048 \
+--hf-access-token $HF_TOKEN \
+--use-chat-template --remove-local-dataset-files \
+--num-generations 2 --batch-size 1024 --num-batches 200 \
+upload-to-hf --hf-repo-id ${HF_REPO_NAME}
+```
+
+When `multi_sampling=True` (Step 3.a), the `--num-generations` parameter should correspond to the desired number of unique completions per prompt. The `--batch-size 1024` argument specifies that 1024 prompts from the input dataset will be processed by the JetStream server for inference. The `--num-batches 200` argument indicates that this process of fetching 1024 prompts, generating completions, and then uploading the resulting prompt-completion pairs will be repeated 200 times. It's important to note that some prompts may be filtered out by pre-processing logic before inference. Additionally, the generated dataset can also be uploaded to Google Cloud Storage (GCS) by setting `upload-to-gcs --gcs-bucket <gcs bucket name> --gcs-data-path <path in gcs bucket>`.
```
Comment (shralex) [inline getting_started/Knowledge_Distillation.md:103]: For some reason this took a couple of reads for me to understand, so I tried to rephrase:

When `multi_sampling=True` (Step 3.a), , the `--num-generations` parameter specifies the number of distinct completions to generate per prompt. The `--batch-size` parameter controls how many prompts are processed per batch, and `--num-batches` defines how many such batches to run. The total number of prompt-completion pairs generated is approximately: `total_pairs = num_batches * batch_size * num_generations`.

For example, with `--batch-size 1024`, `--num-generations 2`, and `--num-batches 200`, this would yield:
`total_pairs = 200 * 1024 * 2 = 409,600` prompt-completion pairs.

Note: Some prompts may be dropped by pre-processing filters, so the final number might be slightly lower.

Then have do a newline, say something about upload-to-hf and then your text about upload to GCS.

File: getting_started/Knowledge_Distillation.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+### 4.a. Fine-tune the Student Model using Dataset Generated in Step 3
+
+```bash
+python3 -m MaxText.sft_trainer MaxText/configs/sft.yml \
+run_name=${RUN_NAME} \
+base_output_directory=${BASE_DIRECTORY}/distillation/deepseek2-16b-distill-llama2-7b \
+tokenizer_path=meta-llama/Llama-2-7b-chat-hf tokenizer_type=huggingface \
+hf_path=${USERNAME_OR_ORG}/${HF_REPO_NAME} train_split='train' train_data_columns=['prompt','completion'] \
+load_parameters_path=${BASE_DIRECTORY}/llama2-7b-chat/scanned/0/items model_name=llama2-7b \
+per_device_batch_size=2 ici_expert_parallelism=-1 ici_fsdp_parallelism=4 \
```
Comment (shralex) [inline getting_started/Knowledge_Distillation.md:117]: does this require a certain hardware or topology ?

File: getting_started/Knowledge_Distillation.md (hunk 1 above)
Comment (shralex) [reply getting_started/Knowledge_Distillation.md:103]: Also can you provide some more detail about pre-processing filters

Comment (SurbhiJainUSC) [reply getting_started/Knowledge_Distillation.md]: Thanks for suggestion, updated!

File: getting_started/Knowledge_Distillation.md (hunk 2 above)
Comment (SurbhiJainUSC) [reply getting_started/Knowledge_Distillation.md:127]: This is just an example command for v4-8. Users would have to modify these sharding configurations accordingly. I have specified in the documentation that this is an example command for v4-8.

**Review by shralex — APPROVED** [review]

File: getting_started/Knowledge_Distillation.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+#### b. Convert Checkpoint to MaxText Format
+MaxText requires checkpoints to be in a specific format. You'll need to convert the downloaded Hugging Face checkpoints to a MaxText-compatible checkpoint.
+
+```bash
+# Get unscanned checkpoint for efficient decoding
+JAX_PLATFORMS=cpu python3 -m MaxText.convert_deepseek_unscanned_ckpt \
+--base_model_path ~/deepseek2-16b-chat \
+--maxtext_model_path ${BASE_DIRECTORY}/deepseek2-16-chat/unscanned \
+--model_size deepseek2-16b
+```
```
Comment (bvandermoon) [inline getting_started/Knowledge_Distillation.md:47]: Is there setup needed for this like `bash setup.sh` (probably on a virtual env)? Please add to the prereqs section if so

**Review by bvandermoon — COMMENTED** [review]
LGTM. Just some nits. Feel free to take ones that make sense and leave off ones that don't

File: getting_started/Knowledge_Distillation.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+Set `multi_sampling` to `True` to generate multiple independent completions per prompt.
+
+
+### 3.b. Generate Dataset using JetStream Server
+In a new tab in your terminal, run the following command to generate dataset from teacher model. Note that this is an example command to run on `v4-8`:
+
+```bash
+python3 -m MaxText.generate_distillation_data \
+--tokenizer-path=deepseek-ai/DeepSeek-V2-Lite-chat \
+--dataset-path HuggingFaceH4/ultrachat_200k --data-split train_sft \
```
[the remark this answers is not in the document]
Comment (SurbhiJainUSC) [reply getting_started/Knowledge_Distillation.md:96]: Sure, I will try to club similar arguments together to reduce number of lines too.

File: getting_started/Knowledge_Distillation.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+--maxtext_model_path ${BASE_DIRECTORY}/deepseek2-16-chat/unscanned \
+--model_size deepseek2-16b
+```
+
+### 2. Obtain and Prepare the Student Model
+
+#### a. Download Model from Hugging Face
+
+```bash
+huggingface-cli login
```
[the remark this answers is not in the document]
Comment (SurbhiJainUSC) [reply getting_started/Knowledge_Distillation.md:54]: This won't be required for the second time, will remove it.

File: getting_started/Knowledge_Distillation.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+```bash
+python3 -m MaxText.generate_distillation_data \
+--tokenizer-path=deepseek-ai/DeepSeek-V2-Lite-chat \
+--dataset-path HuggingFaceH4/ultrachat_200k --data-split train_sft \
+--data-columns messages \
+--max-prefill-length 64 --max-target-length 2048 \
+--hf-access-token $HF_TOKEN \
+--use-chat-template --remove-local-dataset-files \
+--num-generations 2 --batch-size 1024 --num-batches 200 \
```
[the remark this answers is not in the document]
Comment (SurbhiJainUSC) [reply getting_started/Knowledge_Distillation.md]: Fixed!

File: getting_started/Knowledge_Distillation.md (hunk 3 above)
Comment (SurbhiJainUSC) [reply getting_started/Knowledge_Distillation.md]: `setup.sh` is required for first run, right? I have added `pip install -r requirements.txt` step in prereq to update distillation specific dependencies.

Comment (bvandermoon) [reply getting_started/Knowledge_Distillation.md]: Good point. If you are including the command to git clone MaxText though, it should probably still be `bash setup.sh` instead of `pip install -r requirements.txt`. Or it could be better to just link to that first run README and say to run the installation steps under 'Running on Cloud TPUs` through `bash setup.sh` or `pre-commit install`. We should probably split out a separate MaxText prereqs README in the future which basically just has those 3 steps

**Review by bvandermoon — APPROVED** [review]
LGTM pending minor discussion around installation steps

File: requirements.txt [hunk 7]
```diff
@@ -14,6 +14,7 @@ google-cloud-monitoring
 google-api-core
 google-api-python-client
 grain[parquet]>=0.2.6
+huggingface_hub
```
Comment (bvandermoon) [inline requirements.txt:17]: Is this also needed in `requirements_with_jax_stable_stack.txt` and `constraints_gpu.txt`?

**Comment by bvandermoon** [comment]
> LGTM pending minor discussion around installation steps

Also one other comment around the requirements

File: requirements.txt (hunk 7 above)
Comment (SurbhiJainUSC) [reply requirements.txt:17]: Discussed offline to take a note of this to add this in future if Docker/XPK will be supported too.
